### PR TITLE
Benchmark

### PR DESCRIPTION
**Pull request by NicklasHugoy**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,7 +9,7 @@ if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
 endif()
 
 
-set(CMAKE_CXX_FLAGS "-Wall -Wpedantic -Wextra")
+set(CMAKE_CXX_FLAGS "-Wall -Wpedantic -Wextra -g")
```
Comment (NicklasHugoy) [inline CMakeLists.txt:12]: We probably dont want to keep the -g flag

File: src/main.cpp [hunk 2]
```diff
@@ -55,17 +59,22 @@ int main(int argc, char *argv[]){
...
         if(gpu_flag){
             std::cout << "Running analysis using GPU" << std::endl;
-            auto program = parse_to_cfg_transformer<std::set<std::string>>(prog);
-            gpu_analysis(program.nodes);
-            print_digraph_subgraph(program.entryNodes, std::cout, print_result, "main");
+
+            auto program = timeFunc<ScTransformer<std::set<std::string>>>("Creating CFG nodes: ", 
+                parse_to_cfg_transformer<std::set<std::string>>, prog);
+            
+            timeFunc("Total GPU analysis: ", gpu_analysis, program.nodes);
```
Comment (ZcortezZ) [inline src/main.cpp:66]: Kunne sætte det op som et benchmark flag, så timing kun tages hvis det bliver specificeret

File: src/main.cpp [hunk 3]
```diff
@@ -55,17 +59,22 @@ int main(int argc, char *argv[]){
...
+            // print_digraph_subgraph(program.entryNodes, std::cout, print_result, "main");
         }else{
             if(multi_taint_flag){
                 std::cout << "Running multi-taint analysis using CPU" << std::endl;
                 auto program = parse_to_cfg_transformer<SourcedTaintState>(prog);
-                cpu_multi_taint_analysis(program);
+                // cpu_multi_taint_analysis(program);
```
Comment (NicklasHugoy) [inline src/main.cpp:72]: Multi taint analysis probably shouldnt be disabled like this.
I would probably uncomment but just not apply any benchmarking to it

File: src/main.cpp [hunk 4]
```diff
@@ -21,10 +24,11 @@ void print_result(std::set<std::string>& result, std::ostream& stream){
...
 void cpu_analysis(ScTransformer<std::set<std::string>> program){
     TaintAnalyzer analyzer;
-    worklist(program.nodes, analyzer);
 
-    //print_digraph_with_result<std::set<std::string>>(program.nodes, std::cout, print_result);
-    print_digraph_subgraph(program.entryNodes, std::cout, print_result, "main");
+    timeFunc("Analyzing: ", 
+        worklist<std::set<std::string>>, program.nodes, analyzer);
+
+    // print_digraph_subgraph(program.entryNodes, std::cout, print_result, "main");
```
Comment (ZcortezZ) [inline src/main.cpp:31]: Ville være dejligt hvis den stadig ville give et output af analysen.

File: src/matrix_analysis.cu [hunk 5]
```diff
@@ -49,14 +51,22 @@ void gpu_analysis(std::vector<std::shared_ptr<Node<std::set<std::string>>>>& nod
...
     
     Matrix<float> init_state = get_initial_matrix(matrixTransformer.variables.size(), nodes.size());
+    stopwatch.printTimeMicroseconds("Matrix creation ");
+
     auto result_state = analyse(matrixTransformer.matrices, successor_matrix, init_state).to_matrix();
 
+    timeFunc("Move data to cpu: ", 
+        set_node_states, result_state, nodes, matrixTransformer.variables);
     // Set the tainted state on nodes
-    set_node_states(result_state, nodes, matrixTransformer.variables);
```
Comment (NicklasHugoy) [inline src/matrix_analysis.cu:55]: This lil comment became homeless with this change

**Review by NicklasHugoy — COMMENTED** [review]
Minor changes

**Review by ZcortezZ — APPROVED** [review]

**Review by thorulf4 — APPROVED** [review]

**Comment by NicklasHugoy** [comment]
Opdateret med et benchmark argument

**Review by thorulf4 — APPROVED** [review]

**Pull request merged**